PhotoCollection/PhotoCollection/PhotoController.swift

```
@@ -0,0 +1,68 @@
+//
+//  PhotoController.swift
+//  PhotoCollection
+//
+//  Created by Cora Jacobson on 6/10/20.
+//  Copyright © 2020 Cora Jacobson. All rights reserved.
+//
+
+import Foundation
+
+class PhotoController {
+    
+    init() {
+        loadFromPersistentStore()
+    }
+    
+    var photos: [Photo] = []
+    
+    func createPhoto(_ imageData: Data, _ title: String) {
+        let photo = Photo(imageData: imageData, title: title)
+        photos.append(photo)
+        saveToPersistentStore()
+    }
+    
+    func updatePhoto(_ photo: Photo, _ imageData: Data, _ title: String) {
+        if let photoIndex = photos.firstIndex(of: photo) {
+            photos[photoIndex].imageData = imageData
+            photos[photoIndex].title = title
+            saveToPersistentStore()
+        }
+    }
+    
+    // MARK: Persistence
+    
+    private var photoCollectionURL: URL? {
+        let fm = FileManager.default
+        let filename = "PhotoCollection.plist"
+        guard let dir = fm.urls(for: .documentDirectory, in: .userDomainMask).first else { return nil }
+        return dir.appendingPathComponent(filename)
+    }
+    
+    private func saveToPersistentStore() {
+        guard let url = photoCollectionURL else { return }
+        do {
+            let encoder = PropertyListEncoder()
+            let data = try encoder.encode(photos)
+            try data.write(to: url)
+        } catch {
+            NSLog("Was not able to encode data: \(error)")
+        }
```

The way you have this set up is just fine, but I just wanted to show you that you can also throw an error yourself if you wanted to. If you were unable to locate the documents directory (which would cause the computed property above to return nil, you could throw an error that would be caught in the catch block:

``` swift
do {
    guard let url = photoCollectionURL else {
        throw NSError(domain: "Unable to locate documents directory", code: 0)
    }
    let encoder = PropertyListEncoder()
    let data = try encoder.encode(photos)
    try data.write(to: url)
} catch {
    NSLog("Was not able to encode data: \(error)")
}
```